msrestazure/azure_local_creds_prober.py

```diff
@@ -0,0 +1,142 @@
+#--------------------------------------------------------------------------
+#
+# Copyright (c) Microsoft Corporation. All rights reserved.
+#
+# The MIT License (MIT)
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the ""Software""), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+#--------------------------------------------------------------------------
+
+
+import os
+import json
+import platform
+import logging
+from subprocess import PIPE, Popen
+import requests
+from msrest.authentication import BasicTokenAuthentication
+from msrestazure.azure_active_directory import MSIAuthentication, ServicePrincipalCredentials
+
+_LOGGER = logging.getLogger(__name__)
+
+class AzureLocalCredentialProber(object):
+    '''
+    Probing logics:
+    1. Managed service identity
```

Comment:
The progression should be from more specific to less specific. This is the way I'd define the specificity of the different sources of information:

```
Environment variables > local file > local endpoint available. 
```
Following this, does this: 2, 3, 1, 4 make sense?

Reply:
I started with the most secure option, then follow up with SDK home made ones. But you have a point, so I will adjust. Also it is unlikely people would set env variable in a VM with identity